openhands/runtime/impl/local/local_runtime.py

```
@@ -0,0 +1,344 @@
+        self._log_thread.start()
+
+        self.log('info', f'Waiting for server to become ready at {self.api_url}...')
+        self.send_status_message('STATUS$WAITING_FOR_CLIENT')
+
+        await call_sync_from_async(self._wait_until_alive)
+
+        if not self.attach_to_existing:
+            await call_sync_from_async(self.setup_initial_env)
+
+        self.log(
+            'debug',
+            f'Server initialized with plugins: {[plugin.name for plugin in self.plugins]}',
+        )
+        if not self.attach_to_existing:
+            self.send_status_message(' ')
+        self._runtime_initialized = True
+
+    def _find_available_port(self, port_range, max_attempts=5):
+        port = port_range[1]
+        for _ in range(max_attempts):
+            port = find_available_tcp_port(port_range[0], port_range[1])
+            return port
+        return port
+
+    @tenacity.retry(
+        wait=tenacity.wait_exponential(min=1, max=10),
+        stop=tenacity.stop_after_attempt(10) | stop_if_should_exit(),
+        before_sleep=lambda retry_state: logger.debug(
+            f'Waiting for server to be ready... (attempt {retry_state.attempt_number})'
+        ),
+    )
+    def _wait_until_alive(self):
+        """Wait until the server is ready to accept requests."""
+        if self.server_process and self.server_process.poll() is not None:
+            raise RuntimeError('Server process died')
+
+        try:
+            response = self.session.get(f'{self.api_url}/alive')
+            response.raise_for_status()
+            return True
+        except Exception as e:
+            self.log('debug', f'Server not ready yet: {e}')
+            raise
+
+    async def execute_action(self, action: Action) -> Observation:
+        """Execute an action by sending it to the server."""
+        if not self._runtime_initialized:
+            return ErrorObservation('Runtime not initialized')
```

Should we raise an exception, rather than send an ErrorObs? It's not the kind of thing that an agent can fix, and it would end up in its context otherwise